docs(dpg): add introduction to `RequestContext`

archerzz opened the pull request:
1. add introduction section in `README.md` regarding `RequestContext`
2. add sample doc about `RequestContext`

part of #26454

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

archerzz committed:
docs(dpg): add introduction to `RequestContext`

1. add introduction section in `README.md` regarding `RequestContext`
2. add sample doc about `RequestContext`

part of #26454

azure-sdk commented:
**API change check for `Azure.Core`**

API changes are not detected in this pull request for `Azure.Core`

christothes commented on sdk/core/Azure.Core/README.md:
```diff
@@ -82,9 +83,25 @@ SecretClient client = new SecretClient(new Uri("http://example.com"), new Defaul
 
+### Customzing Request Using `RequestContext`
+
+Besides general configuration of _serice client_ through `ClientOptions`, you can customize each individual request sent by _service client_
```
```suggestion
Besides general configuration of _service clients_ through `ClientOptions`, it is possible to customize each individual request sent by a _service client_
```

annelo-msft replied:
You can only do this in protocol methods on a service client, and in select APIs on Generation1 Convenience Clients, where RequestContext is exposed as a parameter.

I'd suggest changing this to "[...] you can customize the request send by the _service client_ using protocol methods or convenience APIs that expose `RequestContext` as a parameter."

archerzz replied:
Done

christothes commented on sdk/core/Azure.Core/README.md:
```diff
@@ -82,9 +83,25 @@ SecretClient client = new SecretClient(new Uri("http://example.com"), new Defaul
+Besides general configuration of _serice client_ through `ClientOptions`, you can customize each individual request sent by _service client_
+using `RequestContext`.
+
+_Service clients_ have methods that can be used to call Azure services. We refer to these client methods _service methods_. You can
+pass `RequestContext` to each _service method_ as a parameter.
```
Suggest combining this sentence with the previous paragraph. 

```suggestion
Each _service client_ method involving a call to an Azure service accepts a `RequestContext` instance as a parameter.
```

annelo-msft replied:
We should remove this, because not all service client methods expose `RequestContext`.

archerzz replied:
I move the sentence regarding _service methods_ back to section "Accessing HTTP Response Details Using ```Response<T>```".

christothes commented on sdk/core/Azure.Core/tests/LowLevelClientTests.cs:
```diff
@@ -183,10 +183,12 @@ public async Task ThrowOnErrorDoesntThrowOnSuccess()
             var mockTransport = new MockTransport(mockResponse);
             PetStoreClient client = CreateClient(mockTransport);
 
+            #region Snippet:SetRequestContext
```
nit: I think it reads more clearly to push RequestContext down to a new line:

```c#
Response response = await client.GetPetAsync("snoopy",
    new RequestContext()
    {
        ErrorOptions = ErrorOptions.Default
    });
```

archerzz replied:
Done. But I removed the snippet directive per comments below.

annelo-msft commented on sdk/core/Azure.Core/README.md:
````diff
@@ -82,9 +83,25 @@ SecretClient client = new SecretClient(new Uri("http://example.com"), new Defaul
+```C# Snippet:SetRequestContext
+Response response = await client.GetPetAsync("snoopy", new RequestContext()
+{
+    ErrorOptions = ErrorOptions.Default
````
I think it's unrealistic that we'd create a new instance of `RequestContext` to set a property to its default value -- could we instead create a new RequestContext and call `AddClassifier(404, isError: false)` on it before passing it to the method?

archerzz replied:
Replaced with a classifier 404 case.

annelo-msft commented on sdk/core/Azure.Core/README.md:
````diff
@@ -82,9 +83,25 @@ SecretClient client = new SecretClient(new Uri("http://example.com"), new Defaul
+});
+```
+
+More on request customization in [RequestContext samples](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/RequestContext.md)
````
This doc is outdated and needs to be updated, specifically this section: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/ProtocolMethods.md#using-requestoptions-to-customize-behavior

archerzz replied:
updated

annelo-msft replied:
Thank you!

annelo-msft commented on sdk/core/Azure.Core/samples/RequestContext.md:
```diff
@@ -0,0 +1,45 @@
+
+## Customize Error Handling
+
+You can suppress exception when error is returned.
```
```suggestion
You can suppress the exception a service method throws when an error response is returned.
```

annelo-msft reviewed (commented):
This is really nice, thank you!  I left a few comments.

archerzz commented on sdk/core/Azure.Core/README.md:
```diff
@@ -22,6 +22,7 @@ you can find the NuGet package [here](https://www.nuget.org/packages/Azure.Core)
 The main shared concepts of Azure.Core (and so Azure SDK libraries using Azure.Core) include:
 
 - Configuring service clients, e.g. configuring retries, logging (`ClientOptions`).
+- Customizing request (`RequestContext`).
```
Done

archerzz committed:
fix sentence in RequestContext.md

Co-authored-by: Anne Thompson <annelo@microsoft.com>

archerzz committed:
update according to comments

- move sentence regarding _service methods_ back to original section "Accessing HTTP Response Details Using ```Response<T>```"
- move `Customizing  request (`RequestContext`).` to a later location
- tune the format of a test case
- add a test case in `LowLevelClientTests.cs` and update the sample on creating `RequestContext`
- update `ProtocolMethods.md`

archerzz commented:
A couple of changes:
- move sentence regarding _service methods_ back to original section "Accessing HTTP Response Details Using ```Response<T>```"
- move `Customizing  request (`RequestContext`).` to a later location
- tune the format of a test case
- add a test case in `LowLevelClientTests.cs` and update the sample on creating `RequestContext`
- update `ProtocolMethods.md`

annelo-msft commented on sdk/core/Azure.Core/samples/RequestContext.md:
```diff
@@ -0,0 +1,45 @@
+
+## Customize Category Of Response By Status Code
+
+You can change the category of response by the returned status code.
```
Lets add one more sentence here saying: "This can be used to treat error responses as success responses in logging and distributed tracing."

archerzz replied:
Done

annelo-msft commented on sdk/core/Azure.Core/tests/LowLevelClientTests.cs:
```diff
@@ -211,6 +212,25 @@ await client.GetPetAsync("snoopy", new RequestContext()
+            var context = new RequestContext();
+            context.AddClassifier(404, isError: false);
+
+            Response response = await client.GetPetAsync("pet1", context);
```
Ah, I see where the issue is now.  We should be able to regenerate or manually update PetStoreClient to bring it up to date with the rest of the generated clients.

archerzz replied:
I've made changes in `PetStoreClient` to be more aligned with a generated client. The test case failure is fixed.

lirenhe commented on sdk/core/Azure.Core/samples/RequestContext.md:
````diff
@@ -0,0 +1,45 @@
+
+```C# Snippet:AddPolicyPerCall
+var context = new RequestContext();
+context.AddPolicy(new AddHeaderPolicy("PerCallHeader", "Value"), HttpPipelinePosition.PerCall);
````
I don't have the context for this. Could you help to explain the behaviour of adding this policy in the context?

archerzz replied:
This will add a policy which will add a header to the request per call. Since it's a `RequestContext`, I assume this is useful in a LRO case in which there are multiple requests. And you can add the header to each request.

lirenhe replied:
I see. is it possible to put the descritpion above the code so reader could know what the below code would do.

archerzz replied:
done

lirenhe commented on sdk/core/Azure.Core/samples/RequestContext.md:
````diff
@@ -0,0 +1,45 @@
+
+```C# Snippet:AddPolicyPerRetry
+var context = new RequestContext();
+context.AddPolicy(new AddHeaderPolicy("PerRetryHeader", "Value"), HttpPipelinePosition.PerRetry);
````
Same above

archerzz replied:
Similar. This will add a policy in each retry request.

JoshLove-msft commented on sdk/core/Azure.Core/samples/ProtocolMethods.md:
````diff
@@ -2,62 +2,62 @@
 
 An API using model types could be:
 
 ```csharp
````
We should move this into a code snippet if possible.

archerzz replied:
The whole document is based on a client which doesn't exists. Since this PR is for adding introduction to `RequestContext`, I try to control the scope of the change. It would be better to solve this in another issue. To be honest, topics in `ProtocolMethods.md` are covered in other documents, maybe we can even delete this.

archerzz replied:
I've filed https://github.com/Azure/azure-sdk-for-net/issues/28315 to further track this issue.

JoshLove-msft commented on sdk/core/Azure.Core/samples/ProtocolMethods.md:
```diff
@@ -2,62 +2,62 @@
-        // }
-        // Response: {}
-        Response SetDog(RequestContent requestBody, RequestOptions options = default);
+// Request: {
```
I'm a little unsure what we are trying to convey with these commented out JSON objects.

archerzz replied:
I guess the original intention is to provide examples of request/response contents.

annelo-msft replied:
@JoshLove-msft, good point.  @archerzz is correct.  In DPG clients, the JSON schema for inputs and outputs is illustrated in doc comments for users to help them know what to put in RequestContent and find in Response.

JoshLove-msft commented on sdk/core/Azure.Core/samples/ProtocolMethods.md:
```diff
@@ -92,65 +92,55 @@ Protocol methods need a JSON object of the shape required by the schema of the s
-        }
-        */
-        client.SetDog(RequestContent.Create(data));
+var data = new {
```
Let's use a code snippet if possible.

archerzz replied:
Tracked in https://github.com/Azure/azure-sdk-for-net/issues/28315

JoshLove-msft commented on sdk/core/Azure.Core/samples/ProtocolMethods.md:
```diff
@@ -92,65 +92,55 @@ Protocol methods need a JSON object of the shape required by the schema of the s
+## Using `RequestContext` to customize behavior
 
-`RequestOptions` has some advanced features which allow users to customize behavior:
+`RequestContext` has some advanced features which allow users to customize behavior of each individual request.
```
```suggestion
`RequestContext` has some advanced features which allow users to customize the behavior of each individual request.
```

JoshLove-msft commented on sdk/core/Azure.Core/samples/ProtocolMethods.md:
```diff
@@ -92,65 +92,55 @@ Protocol methods need a JSON object of the shape required by the schema of the s
-                name = "Buddy"
-        }));
-        var content = response.Content;
+Response response = client.GetDog(RequestContent.Create(new {
```
I assume we are using an anonymous type and relying on the default STJ serializer to serialize to JSON? Can we explain that because I don't think it is obvious from the usage alone.

archerzz replied:
explanation added

JoshLove-msft commented on sdk/core/Azure.Core/samples/ProtocolMethods.md:
````diff
@@ -92,65 +92,55 @@ Protocol methods need a JSON object of the shape required by the schema of the s
 ```
 
-Protocol methods, just like other methods that use models, throw a `RequestFailedException` when an error code is returned. This default behavior can be changed using `RequestOptions` discussed below.
+Protocol methods, just like other methods that use models, throw a `RequestFailedException` when an error code is returned. This default behavior can be changed using `RequestContext` discussed below.
````
```suggestion
Protocol methods, just like other methods that communicate with Azure services, throw a `RequestFailedException` when an error code is returned. This default behavior can be changed using `RequestContext` discussed below.
```

archerzz committed:
fix(test): PetStoreClient not support external classifier

1. remove `ResponseClassifier200`, use `StatusCodeClassifier`
2. change logic of `GetPet[Async]` to align with generated client

archerzz committed:
Tune the words in `ProtocolMethods.md`

Co-authored-by: JoshLove-msft <54595583+JoshLove-msft@users.noreply.github.com>

archerzz committed:
Fix and tune words in `ProtocolMethods.md`

Co-authored-by: JoshLove-msft <54595583+JoshLove-msft@users.noreply.github.com>

archerzz committed:
add explanation of anonymous class serialization

archerzz commented:
A couple of changes:
- add explanation of default JSON serialization by `System.Text.JSON` through runtime reflection
- fix and tune some words
- fix the test case regarding classifier by changing the codes of `PetStoreClient` to be more aligned with generated client.

archerzz committed:
Update RequestContext.md

annelo-msft commented on sdk/core/Azure.Core/README.md:
```diff
@@ -183,6 +184,20 @@ Console.WriteLine(value.ScheduledPurgeDate);
 
+### Customzing Request Using `RequestContext`
+
+Besides general configuration of _serice clients_ through `ClientOptions`, it is possible to customize the requests sent by _service clients_
```
```suggestion
Besides general configuration of _service clients_ through `ClientOptions`, it is possible to customize the requests sent by _service clients_
```

annelo-msft reviewed (approved):
This is good enough to merge at this point.  We still need to come back and review these docs holistically as part of https://github.com/azure/azure-sdk-for-net/issues/27770

archerzz committed:
Update sdk/core/Azure.Core/README.md

Co-authored-by: Anne Thompson <annelo@microsoft.com>

archerzz committed:
Update RequestContext.md

lirenhe reviewed (approved):
LGTM
